crates/matrix-sdk/src/room_directory_search.rs

```diff
@@ -0,0 +1,370 @@
+            if !body.get_field::<String>("since").is_ok_and(|s| s == self.next_token) {
+                return false;
+            }
+
+            if !body.get_field::<u32>("limit").is_ok_and(|s| s == Some(self.limit)) {
+                return false;
+            }
+
+            // The body's `filter` field has `generic_search_term` equal to the matcher's
+            // next_token.
+            if !body.get_field::<Filter>("filter").is_ok_and(|s| {
+                if self.filter_term.is_none() {
+                    s.is_none() || s.is_some_and(|s| s.generic_search_term.is_none())
+                } else {
+                    s.is_some_and(|s| s.generic_search_term == self.filter_term)
+                }
+            }) {
+                return false;
+            }
+
+            method("POST").matches(request)
+                && path_regex("/_matrix/client/../publicRooms").matches(request)
+        }
+    }
+
+    fn get_first_page_description() -> RoomDescription {
+        RoomDescription {
+            room_id: RoomId::parse("!ol19s:bleecker.street").unwrap(),
+            name: Some("CHEESE".into()),
+            topic: Some("Tasty tasty cheese".into()),
+            alias: None,
+            avatar_url: Some("mxc://bleeker.street/CHEDDARandBRIE".into()),
+            join_rule: ruma::directory::PublicRoomJoinRule::Public,
+            is_world_readable: true,
+            joined_members: 37,
+        }
+    }
+
+    fn get_second_page_description() -> RoomDescription {
+        RoomDescription {
+            room_id: RoomId::parse("!ca18r:bleecker.street").unwrap(),
+            name: Some("PEAR".into()),
+            topic: Some("Tasty tasty pear".into()),
+            alias: RoomAliasId::parse("#murrays:pear.bar").ok(),
+            avatar_url: Some("mxc://bleeker.street/pear".into()),
+            join_rule: ruma::directory::PublicRoomJoinRule::Knock,
+            is_world_readable: false,
+            joined_members: 20,
+        }
+    }
+
+    async fn new_server_and_client() -> (MockServer, Client) {
+        let server = MockServer::start().await;
+        let client = logged_in_client(Some(server.uri())).await;
+        (server, client)
+    }
+
+    #[async_test]
+    async fn search_success() {
+        let (server, client) = new_server_and_client().await;
+
+        let mut room_directory_search = RoomDirectorySearch::new(client);
+        Mock::given(RoomDirectorySearchMatcher { next_token: None, filter_term: None, limit: 1 })
+            .respond_with(ResponseTemplate::new(200).set_body_json(&*test_json::PUBLIC_ROOMS))
+            .mount(&server)
+            .await;
+
+        room_directory_search.search(None, 1).await.unwrap();
+        let (results, _) = room_directory_search.results();
+        assert_eq!(results.len(), 1);
+        assert_eq!(results[0], get_first_page_description());
+        assert!(!room_directory_search.is_at_last_page);
```

Comment:
```suggestion
        assert!(!room_directory_search.is_at_last_page());
```

Test the public API as much as possible, except if you need to test internal stuff.